sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -246,304 +247,620 @@ abstract class RapidsShuffleThreadedWriterBase[K, V](
+                    if (bytesRead > 0) {
+                      os.write(buffer, 0, bytesRead)
+                      remainingBytes -= bytesRead
+                    } else {
+                      throw new IOException(
+                        s"Unexpected end of file reading partition $partitionId")
+                    }
+                  }
+                }
+            }
+          }
+        }
+      }
+    } finally {
+      // Close all partial file streams
+      partialFileInputs.foreach { in =>
+        try {
+          in.close()
+        } catch {
+          case _: Exception => /* ignore */
+        }
+      }
+      partialFiles.foreach { pf =>
+        try {
+          pf.file.delete()
+        } catch {
+          case _: Exception => /* ignore */
         }
       }
-      writeMetrics.incWriteTime(System.nanoTime() - writeStartTime)
     }
-    commitAllPartitions(mapOutputWriter, false /*non-empty checksums*/)
-  }
 
-  // taken from BypassMergeSortShuffleWriter
-  // this code originally called into guava.Closeables.close
-  // and had logic to silence exceptions thrown while copying
-  // I am ignoring this for now.
-  def writePartitionedDataWithStream(file: java.io.File, writer: ShufflePartitionWriter): Unit = {
-    withResource(new FileInputStream(file)) { in =>
-      withResource(writer.openStream()) { os =>
-        Utils.copyStream(in, os, false, false)
+    writeMetrics.incWriteTime(System.nanoTime() - mergeStartTime)
+
+    // Clean up partial files
+    partialFiles.foreach { partialFile =>
+      if (partialFile.file != null && partialFile.file.exists()) {
+        partialFile.file.delete()
       }
     }
-  }
 
-  // taken from BypassMergeSortShuffleWriter
-  // this code originally called into guava.Closeables.close
-  // and had logic to silence exceptions thrown while copying
-  // I am ignoring this for now.
-  def writePartitionedDataWithChannel(
-    file: File,
-    outputChannel: WritableByteChannelWrapper): Unit = {
-    // note outputChannel.close() doesn't actually close it.
-    // The call is there to record keep the partition lengths
-    // after the serialization completes.
-    withResource(outputChannel) { _ =>
-      withResource(new FileInputStream(file)) { in =>
-        withResource(in.getChannel) { inputChannel =>
-          Utils.copyFileStreamNIO(
-            inputChannel, outputChannel.channel, 0L, inputChannel.size)
-        }
+    // Commit final merged output
+    commitAllPartitions(finalWriter, true)
+  }
+  
+  /**
+   * Extract outputTempFile and partitionLengths from ShuffleMapOutputWriter.
+   * 
+   * IMPORTANT: For partial batches in multi-batch scenario, we should NOT call
+   * commitAllPartitions() as it would rename/move outputTempFile. Instead, we just
+   * close the streams and extract the file reference for later merge.
+   * 
+   * NOTE: Uses reflection to access LocalDiskShuffleMapOutputWriter internals.
```

Comment:
What if it is not that? Have we verified that it will be a LocalDiskShuffleMapOutputWriter before we ever go down this optimization path? I see the exception below that tells us maybe we should write our own. I don't really feel comfortable with putting this in production without some precautions to avoid crashing a user's query.